### ISSUE TITLE
Pay for what you use: only matching captures or greedily match first.

### ISSUE THREAD

**Issue by V0ldek** [rust-lang/regex#707]
Hello.

It would be nice to have a feature that sits somewhere in between `find` and `captures`. Often I don't care about the groups that were unmatched, I want the groups that _did_ match. An extreme example of this would be a tokenizer implementation, where you would put different possible matches as alternate capture groups, get a match and using the matched group determine the type of the token. So imagine a regex like:
```rust
Regex::new(r#"(?x)
    (if)   |  # IF token
    (else) |  # ELSE token
    (for)  |  # FOR token
    [a-zA-Z][a-zA-Z_0-9]*  # IDENTIFIER token
"#);
```
Any string starting with a character you throw at this is going to match at most two groups and the desired behavior is to treat `"if"` as an IF token, not an IDENTIFIER. In general, imagine dozens of clauses with an assumption that I only care for the left-most matched group. The way it would work now is I would have to call `captures()` and iterate over all the captures discarding `None`s until I find the first matched group (skipping group zero obviously). This is wasteful if there are many non-matched groups, so ideally you'd want to just get the first matched group for free, without going through all the clauses. This adds up if the haystack is long and many matches are performed against it, since the iterate-over-all-groups step has to be done every time.

Ideally, the API would allow us to ask only for the matched groups. In a case where there are a lot of groups but many of them are mutually exclusive (or more generally a lot of the groups fail to match often) it would be much better than having to iterate over all the groups. 

In the particular use-case I presented it would be enough to just hold the smallest index of a matched group, but that seems much more specific and less useful for an "average user". Keeping the groups ordered from left to right would be expected, however.

In other words, sometimes we might want to ignore non-matched groups and in the spirit of "Pay for what you use" it'd be nice if the API provided us with such a function.

**Comment by ethanpailes** [comment]
@V0ldek I'm not sure that this API would actually provide any performance benefits. Internally, the regex engine must set up a pair of slots to track each capture group ahead of time (one slot for the beginning of the group and one for the end). To provide an API that only returns the matched capture groups, we would still need to iterate over all of the slots. I think the best thing for you to do would probably be to implement a convenience wrapper around the existing regex API.

**Comment by V0ldek** [comment]
@ethanpailes 

I'm not familiar with the codebase, but from your answer I draw two assumptions:

1. The engine is very dependent on tracking all the slots ahead of time and changing that is impossible/way too involved to be worth it.
2. The order in which slots are set to `Some` during a match is unpredictable.

If those assumptions are correct then I'd say you're right. The only thing that seems feasible under the above is a list of all captured groups but unordered (just add them to a list when their slot is set to `Some`). But that isn't very useful and sorting them by indices would way outweigh any gains from skipping the `None`s. Can you relate to the above two points?

**Comment by BurntSushi** [comment]
@ethanpailes is right and I see no way to work around it. Not even theoretically.

If checking the capturing groups is too wasteful for your use case, then I'd suggest hand-rolling your tokenizer or using a more purpose built tool for tokenizing.
